### PR TITLE
Use global grid size if grid size is omitted in the preset

### PR DESCRIPTION
**Pull request by Karolis-Sakavicius**
* Use global grid size if grid size is not specified in a preset
* Allow negative coordinates to be used

---

Has been discussed in https://github.com/gTile/gTile/issues/264, but without float-style functionality yet.

Couple things:
1.  I tried to introduce some kind of preview functionality in prefs (e.g. a "Preview" button) that would show the user how current preset will look like (the same way we show where the window is gonna be put while using the UI). But it looks like I can't access `ui.main` in prefs, neither `St`. :disappointed: Maybe somebody knows some ways to work around this obstacle?
2. How do I run the tests? We have `tilespec.spec.ts` but `mocha` is complaining about ES6 syntax.

### CODE REVIEW

**Comment by scherepanov** [comment]
Mmmm... may be we can use current grid sizes instead of introducing global default grid size? What is your use case?
Let me think a little bit what is more intuitive behavior.
Great for allowing negative numbers!
Did you buy an idea on using floats in range 0.0 - 1.0, for specifying window position on any grid sizes? Any thoughts?

**Comment by Karolis-Sakavicius** [comment]
@scherepanov I'm not introducing a new grid size, global grid size in this case refers to whatever is selected in the UI or through the keyboard shortcut. My use case is as follows:
I want to have a 6x3 keyboard, which will be responsible for tiling and in order for me to move the window to some specific location on the grid, I'd use a "global change grid size" accelerator or a hotkey which changes grid to some specific one, then I'd click a single button on this 6x3 and it will move the window to that location, relative to the grid I just switched to. Basically first key is always 0:0, second is 1:0 and so on in this arrangement:
```
+-----+-----+-----+-----+-----+-----+-----+
| 0:0 | 1:0 | 2:0 | 3:0 | 4:0 | 5:0 | 6:0 |
+-----+-----+-----+-----+-----+-----+-----+
| 0:1 | 1:1 | 2:1 | 3:1 | 4:1 | 5:1 | 6:1 |
+-----+-----+-----+-----+-----+-----+-----+
| 0:2 | 1:2 | 2:2 | 3:2 | 4:2 | 5:2 | 6:2 |
+-----+-----+-----+-----+-----+-----+-----+
```

This is the same reason I wanted to get some visual feedback when switching to the other grid size. :)

**Comment by scherepanov** [comment]
@Karolis-Sakavicius thanks for explanations.

Thanks for explanation, everything makes sense.

I think we need to change tile definition. With introduction of negative time position, we should be able to define tile position relative to right-down corner, using negative number. Positive numbers used to define relative position to left-upper corner.

0:0 define left upper corner. For right down corner I need to say -0:-0, which is same as 0:0. 

With introduction of negative numbers, we need to put tile coordinates from 1 instead of 0. Left upper corner will became 1:1 instead of 0:0. That will allow to specify -1:-1 as lower left corner.

Let me know if you have other ideas how to handle it.

For floats, I think we need both options - snap to grid and absolute. Though I cannot think of a good syntax to distinguish between both cases. In my opinion, users would expect from gTile to snap windows to grid edges, let's stick to this approach. If you can suggest syntax to distinguish between snap and absolute windows position, let's do it.


I had some ideas how to make gTile autotiling capabilities expandable by users. We need to allow user to define code snippet, and invoke it with keyboard shortcut (or menu).
Code snippet will receive as parameter 
-Current grid size
-list of windows, with their current position on screen.
- Current window with focus
- Previous window with focus
- Settings for keyboard shortcut (like 2:2 4:4 6x5, number of columns or rows to arrange windows to, etc)
Code snippet will return list of windows that need to be moved/resized, along with new sizes, either in tile syntax, or in absolute position numbers.

That would allow advanced users to implement any tiling they want.
I am often getting some "rather strange" in my opinion autotiling requests. Though all of them has rights to exist and will find it's audience.
With ability to export/import settings, autotiling code snippets can be shared in community.

Question: Do you know how it is possible to dynamically load JS from string and execute it, passing/receiving parameter?

**Comment by Karolis-Sakavicius** [comment]
@scherepanov 
```
0:0 define left upper corner. For right down corner I need to say -0:-0, which is same as 0:0.
```
The idea sounds good to me and would probably be more intuitive than the current approach but I'm a bit worried about backwards compatibility of this change. What is the gTile release cycle? Do we have some major/minor versions, so we could clearly communicate to our users that some changes were introduced that are not backwards compatible? Should I mark this change somehow when I implement it?
Also, I suppose, 0:y and x:0 would be invalid input now, right?

Regarding floats:
I personally would probably expect those to not snap by default, essentially meaning that I have 2 options here - either I set the presets using grid logic, where I supply some grid and place the windows to some specific tiles or I use percentages of the screen. 0.0:0.0 0.25:0.25 would be "place window at the top left corner and expand to 25% of the width/height of the screen". Quite similar to CSS in a way.
If we want to approximate we could probably introduce something like `~0.0:0.0 0.25:0.25`, meaning that gTile will try to snap it to the nearest grid line, either supplied as a first argument (`3x3 ~0.0:0.0 0.25:0.25`) or to the globally set grid if grid is not explicitly set in the preset.
What do you think?

Regarding autotiling: What I'm mostly concerned about autotile solutions is specific windows. For example, when I start my workday, first thing I do is put windows to their respective places (firefox on the left, bunch of terminals on the right, IDE in the middle, etc) and I think lots of people do the same. But with autotiling, how do I know which window will go where? I certainly don't want some docker log terminal that I never look at to take 1/3rd of my screen and firefox to be tiny.
Having some syntax to set some rules for respective windows, by their title for example, would be totally amazing! But we either need to create a UI for that or come up with some reasonable syntax. Both solutions are pretty hard. :disappointed: 

`Question: Do you know how it is possible to dynamically load JS from string and execute it, passing/receiving parameter?`
We could probably leverage `eval()`, but I'm pretty weary on that, as some arbitrary code can crash the gnome-shell. I certainly crashed mine while working on gTile. Maybe if we present it as a "very advanced" feature with a big warning that it might crash their desktop. :thinking: If we had something easier and safer to use alongside, it would be amazing for the users I think.

**Comment by scherepanov** [comment]
Great! Everything makes sense.
Your original question - how to run tests, I do not remember. May be with bazel run?
New tiling syntax, with negatives and floats:
1) I think we should switch to new syntax where 0:0 is not a valid entry.
2) settings file should be modified to use new syntax (increase all tile number by 1).
3) New compiled binary settings file will be included in distribution. That will ensure compatibility of new version.
4) Float is in range of 0.0 - 1.0
5) Great idea to add ~ to tile syntax for float! By default float is absolute. If preceded by tilda ~, gTile will snap to nearest grid line.
6) Everything is mix and match - integer - positive/negatives, absolute and snapped floats can be used together.
7) Omitted grid size means use current grid size (you call it global grid size, I think correct would be to call current grid size).
8) You can partially omit tile definition. Incomplete tiling definition means that user input is expected to finalize window placement. On invoking incomplete tiling definition, window will be placed as much as possible to new position, and GUI will pop-up to allow user to complete placement. Example of incomplete tiling definition is 1: 2:5. In this case, window should be moved to left to vertical line 1, right down corner placed to 2:5. Unspecifid sizes of window shoudl be kept. It will allow user to set height of window.
8) I think we can implement "screen config" for placing many windows in predefined position. You can specify title mask, and window position. On invoking, gTile will go through list of all opened window, read title, check for matching mask, and if match place in predefined position. You can specify several positions for same mask, for case when you need to open several windows for same program (I have several terminals always on screen, always in same position). Window will be placed in next position in a list.
9) I do not think gTile should go into business of starting applications. That is way out of gTile scope. We only do window position management, but doing it better than anybody else. Any desktop has option to start set list of applications on login. We will supplement this with placing all started applications to predefined positions.
10) gTile badly need easily customizeable autotiling for whole screen. All your arguments are completely valid, but anyway users will use it a lot. Yeah, customizeable code will easily crash gnome-shell. This is why it will be for very advanced users only. With mechanism to easily share customizeable autotiling, that would be a hit with users.

May be, autotiling should be moved to separate ticket.

**Comment by Karolis-Sakavicius** [comment]
Sounds good! I'll try to address the syntax things here and autotiling should be separate.:)

**Comment by scherepanov** [comment]
@Karolis-Sakavicius is that all you want to have in this PR, or you are planning to do more change?

**Comment by Karolis-Sakavicius** [comment]
@scherepanov No no, there will be more :) I'm short on time, as always.

**Comment by Karolis-Sakavicius** [comment]
Ok, this should be fairly testable. Will fix the merge conflicts later. Fractional presets should work here, as well as "snapping to grid" behavior.

Fractional presets are from 0.0 till 1.0 ("1" represents a tile, "1.0" represents 100%). Tilde (~) before a coordinate pair will show that this pair should be snapped to the current or specified grid.
E.g. `3x3 ~0.3:0.3 ~0.6:0.6` top left corner as well as bottom right will be snapped to the grid. `0.3:0.3 0.6:0.6` are absolute.
Negative coords are also allowed.
All coordinates now start from 1, instead of 0.

What's left:
* Reintroduce preset validations.
* Update README.

@scherepanov whenever you have some free time, could you try to play around with this?
Regarding partial tiling definitions - that turns out is quite difficult to do. `showTiling` is not very flexible and there is no way to highlight the current window position in tiles as of now. Also, it's difficult to use some non-standard grid there (that is not present in the grid sizes array but is specified in the preset), and it's quite unclear how to handle cases where grid size is not present at all (partial tiling but with fractional values without tildes ~). Maybe let's keep it to some other time in the future.

Also bazel's cyclic dependencies validation makes it very hard to incrementally refactor things. :disappointed: Maybe that's why we have almost all the logic in a single file...

**Comment by scherepanov** [comment]
Great!
I agree implementing partial tiling definition would require significant redesign. Let’s keep it for future.
Will play with this PR, see how it works.

**Comment by scherepanov** [comment]
BTW any reasons why fractional coordinates implemented for pair?
I think we should allow mix and match of integer, fractional, fractional with snap (let’s call them snap), and negative integer.
All below should be valid:
2:2 -2:-2      Window is taking center of screen with one tile border around
0.3:~0.3 ~0.6:-1.    Left is fractional 0.3 and snap 0.3, right is snap 0.6 and lower edge of screen


I think it is not hard to support mix and match of any coordinate types.

**Comment by scherepanov** [comment]
PR works as expected.
I found problem.
Added log with window new X:Y W:h, function moveWindowToRect, line 1385:
    log(`moveWindowToRect x:y ${rect.origin.x}:${rect.origin.y} w:h ${rect.size.width}:${rect.size.height}`)

I logs float numbers: for fractional tile definition:
Jan 16 11:12:58 chi-fed26-scherepanov gnome-shell[473728]: gTile Preset resize 3  is 4x4 0.3:0.3 0.6:0.6, 0.6:0.6 -1:-1
Jan 16 11:12:58 chi-fed26-scherepanov gnome-shell[473728]: gTile moveWindowToRect x:y 1152:681.6 w:h 1152:633.6

For start, it will make neighboring windows to overlap by 1 pixel. 
New coordinates should be integer, and neighboring window with same edge coordinates should not overlap or have gap between them.
We need to calculate edge coordinate, and round it. It is a correct calculation for left/upper corner.
For right/down corner, we need to subtract 1 from round result. That will make sure that two windows do not overlap by 1 pixel.
And, we should not allow a single pixel to slip between windows. That, for example, will create a bright white single-pixel line, if both windows has dark theme, and background is bright. I had user who loudly complained about this, took some time to figure it out.
And, 1.0 should be width of screen -1 pixel. i.e. right edge of window should be last pixel on screen.

Related problem: due to advance calculation of tile width/height in int, there is a very pronounced error when you have large odd tile size.
For example, set grid size to 97:97 (max grid size is 100x100, 97 is largest prime udder 100)
When you will be placing window to rightmost position, you will see a large gap between screen edge and window right edge.
That is related problem, and we need to have it in mind when fixing this problem.

**Comment by Karolis-Sakavicius** [comment]
Reworked the thing, to allow approximations by either x or y. Now validations and parsing is a bit easier to understand.
Also reinstated the validations & adapted those to the new syntax.

Regarding rounding errors - great point! Maybe if we always rounded left & top edges up and bottom & right edges down, it would solve the problem? I can't come up with an example where it would fail. And it should not produce rounding errors when the window is all the way to the right or all the way to the bottom, as these would always return an exact integer value.

I'll try to play around.

**Comment by scherepanov** [comment]
Rounding all coordinated down will produce overlapping by 1 pixel for neighboring windows

**Comment by Karolis-Sakavicius** [comment]
Imagine 2 windows side by side.
Right edge of the left window resolves to x=1000.66 (it gets rounded **down** to 1000).
Left edge of the right window resolves to the same x = 1000.66 (it gets rounded **up** to 1001).
I think that should solve the problem.

I'm a bit concerned if I'm missing something at the very edges of the screen. 
Top/left should not be a problem as we are multiplying by zero there: 
```
            const positiveTileNumber = this._convertNegativeCoord(this.gridWidth, this.luc.xy.x);
            left = Math.floor(workArea.origin.x + (positiveTileNumber - 1) * elemSize.width);
```
so it will always result in `workArea.origin.x`.

Bottom right is a different story, as it might produce a 1px bigger value than the screen size, right? Should be enough to just look if it's bigger and if it is, we can just use the max edge size, so it does not bleed to the other screen.

I'll push some "work in progress" stuff regarding this, if you have some time to play around as well.

**Comment by scherepanov** [comment]
Case:
Window edge resolved exactly to 1000
Rounding down and up will yield same 1000

**Comment by scherepanov** [comment]
IMHO right algo would be:
Round value, and use it for left/upper corners
Round value and subtract 1, and use it for right/down corner

**Comment by Karolis-Sakavicius** [comment]
Seems to work fine, but it leaves 1px margins on the right side, as well as on the bottom of the work area. 
This is not really visible (at least on my screen) but has a very ugly side effect - when pointer reaches the rightmost pixel of the screen, it changes from a default pointer to a "resize one". This is awful if one wants to use scrollbar to scroll as you can't just yank your mouse to the right side and grab the scrollbar, you have to be precise.

I know that it is caused by that 1px margin because if I navigate my window to the right and let ubuntu snap it to place, the problem disappears.

I can't come up with anything better apart from using rounds, as you suggested, but also checking if it's the edge of the screen, and if it is - add the 1px back. A bit ugly, but I can't come up with anything more intuitive.

**Comment by scherepanov** [comment]
It is very strange.
For example, if screen width is 1024, multiplying on 1.0 would still give 1024, subtracting 1 will make it 1023.
Screen pixel X coordinates are from 0 to 1023, inclusive.
1023 would be a correct result for window placement.
Based on your description, you are placing window to X 1022, which leaves pixel 1023 out of window with all ugly effects.
Can you add log of windows position, like I suggested. I want to check what is going on.

**Comment by Karolis-Sakavicius** [comment]
@scherepanov I pushed the code, your calculations make perfect sense. Now I'm thinking that it might be due to the window bleeding to another monitor. I can't see it because I don't have another monitor.
Could you check if the current implementation bleeds to another monitor when you tile the window on the left monitor?

Could it be related to that weird "tiled" property in gnome-shell that we can't set? When default shortcut (super+arrow right) is used, the window is marked as "tiled" and hence it behaves in a different way - not showing this resize thingy when mouse reaches the right edge.

I'll add the logs soon.

**Comment by Karolis-Sakavicius** [comment]
Preset: `97x97 1:1 97:97`
Log: `gnome-shell[2695]: gTile moveWindowToRect x:y 0:27 w:h 3440:1380`

**Comment by scherepanov** [comment]
My guess your monitor size is 3440:1380.
It means, there is a single line bleeds outside screen right and down.
Try to modify calculation like I suggested - round number, use it for left/upper corner. For right/down corner, add 1 to rounded number.

**Comment by Karolis-Sakavicius** [comment]
Pushed. Guess we will be able to close this https://github.com/gTile/gTile/issues/236 after this PR is merged, even though it introduces this weird resize effect that I don't like.
I'm pretty sure it's related to that `tiled` state, described here: https://github.com/gTile/gTile/issues/104 . Would probably make sense to revisit it to see if something changed and we can now set this. Maybe you have looked into it recently?

Please test the current implementation and let me know if you find something. I'm aware of the issue with margins on maximized screens - will address that some other day and add another commit. As well as READMEs and explanations how to use the new feature.

**Comment by scherepanov** [comment]
Looks good.
Calculations are correct. I do not see line or row of pixels leaking to other monitor.
All new syntax with mix and match works correctly.
I cannot reproduce problem you describing.
Please do not forget to fix README - it would be very confusing for users to use new code.

File: preset_parser.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { XY, TileSpec, GridSize, TupleHolder, MAX_TUPLE_MEMBER_VALUE } from './tilespec'
+
+/**
+ * parsePreset parses a sequence of TileSpec objects from a string like "8x8 0:0
+ * 0:7, 8x10 0:0 2:7" or "8x8 0:0 0:7, 0:0 2:7"
```
Comment (gonzojive) [inline preset_parser.ts:5]: This comment is out of date now that 0 is an invalid coordinate, right?

File: preset_parser.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+function hasImpliedGridSize(singlePreset: string): boolean {
+    return singlePreset.trim().split(" ").length === 2;
+}
+
+/**
+ * Parses a value like like 6x4 or 1:2 into {X: 6, Y: 4} or {X: 1, Y: 2}.
+ */
+function parseTuple(unparsed: string, delim: string) {
```
Comment (gonzojive) [inline preset_parser.ts:51]: Writing an explicit return type makes functions more readable. Please add ": XY"

File: preset_parser.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    // parsing grid size in unparsed XdelimY, like 6x4 or 1:2
+    const gssk = unparsed.split(delim);
```
Comment (gonzojive) [inline preset_parser.ts:53]: What is ggsk? Please use a more verbose variable name.

File: tilespec.spec.ts [hunk 4]
```diff
@@ -60,7 +60,6 @@ describe("TileSpec.isFullscreen", function() {
 
 describe("TileSpec.parsePreset error cases", function() {
     const badPresets = [
-        '0:0 1:1',
```
Comment (gonzojive) [inline tilespec.spec.ts:63]: Why is this no longer bad?

File: tilespec.ts [hunk 5]
```diff
@@ -1,54 +1,213 @@
-const MAX_TUPLE_MEMBER_VALUE = Number.MAX_SAFE_INTEGER;
-
 /**
  * TileSpec represents a rectangular area on display by means of specifying a
  * number of evenly spaced tiles and two corners.
  */
```
Comment (gonzojive) [inline tilespec.ts:4]: This type is now a bit more complicated. May you please update the docstring?

File: tilespec.ts [hunk 6]
```diff
@@ -1,54 +1,213 @@
...
 export class TileSpec {
     readonly gridWidth: number;
```
Comment (gonzojive) [inline tilespec.ts:6]: Is grid width and height no longer required? If so, I would make these number|null.

File: tilespec.ts [hunk 7]
```diff
@@ -1,54 +1,213 @@
...
     readonly gridHeight: number;
-    readonly luc: XY;
-    readonly rdc: XY;
+    readonly luc: TupleHolder;
```
Comment (gonzojive) [inline tilespec.ts:8]: We should probably rename these variables to leftUpper and rightLower or something.

File: tilespec.ts [hunk 8]
```diff
@@ -1,54 +1,213 @@
...
+     * E.g. -1:-1 on a 3:3 grid is a 3:3, as well as -1:3.
+     */
+    _convertNegativeCoord(gridEdges: number, coord: number) {
+        if (coord < 0) {
+            return gridEdges + coord + 1;
+        } else {
+            return coord;
+        }
+    }
+}
```
Comment (gonzojive) [inline tilespec.ts:103]: Comment?

File: tilespec.ts [hunk 9]
```diff
@@ -1,54 +1,213 @@
...
+ *
+ * E.g. ~0.75:0.75 is {X:0.75,Y:0.75}, types - 'percentage' & 'percentage'
+ * approximate - true.
+ */
+export class TupleHolder {
+    raw: string;
+    xy: XY;
+    types: CoordinateTypesHolder;
+
+    constructor(raw: string) {
```
Comment (gonzojive) [inline tilespec.ts:119]: Needs a comment fully explaining the syntax.

File: tilespec.ts [hunk 10]
```diff
@@ -1,54 +1,213 @@
...
+        this.raw = raw;
+
+        const gssk = this.raw.split(':');
```
Comment (gonzojive) [inline tilespec.ts:122]: See other comment on gssk as a variable name.

File: tilespec.ts [hunk 11]
```diff
@@ -1,54 +1,213 @@
...
+export const MAX_TUPLE_MEMBER_VALUE = Number.MAX_SAFE_INTEGER;
+
+/**
+ * Tuple Holder represents a single starting or ending point (x and y coordinates),
+ * as well as the type of the coordinate - "tile", "approx_percentage" or "percentage" now.
+ *
+ * E.g. ~0.75:0.75 is {X:0.75,Y:0.75}, types - 'percentage' & 'percentage'
+ * approximate - true.
+ */
+export class TupleHolder {
```
Comment (gonzojive) [inline tilespec.ts:114]: I recommend a more succinct name such as "CornerSpec," which is more consistent with the naming of the existing "TileSpec" type.

File: tilespec.ts [hunk 12]
```diff
@@ -1,54 +1,213 @@
...
+        this._validateCoordinate(gssk[1]);
+    }
+
+    /**
+     * Allowed values:
+     * 1.0 (exact match)
+     * Any float from 0.0 till 0.999..., with or without preceding approx indicator (~)
+     * Any positive or negative integer, except 0
+     */
+    _validateCoordinate(coord: string) {
```
Comment (gonzojive) [inline tilespec.ts:177]: Here and above, prefer an explicit return type for readability.

File: tilespec.ts [hunk 13]
```diff
@@ -1,54 +1,213 @@
...
+
+        this.xy = this._parseTuple(gssk);
+        this.types = this._parseTypes(gssk);
+    }
+
+    toString() {
+        return this.raw;
+    }
+
+    _parseTuple(tuple: Array<string>) {
```
Comment (gonzojive) [inline tilespec.ts:134]: Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style

File: tilespec.ts [hunk 14]
```diff
@@ -1,54 +1,213 @@
...
+        const x = this._parseCoordinate(tuple[0]);
+        const y = this._parseCoordinate(tuple[1]);
+
+        return new XY(x, y);
+    }
+
+    _parseTypes(tuple: Array<string>) {
```
Comment (gonzojive) [inline tilespec.ts:141]: Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style

File: tilespec.ts [hunk 15]
```diff
@@ -1,54 +1,213 @@
...
+        const typeX = this._parseType(tuple[0]);
+        const typeY = this._parseType(tuple[1]);
+
+        return new CoordinateTypesHolder(typeX, typeY);
+    }
+
+    _parseCoordinate(coord: string) {
```
Comment (gonzojive) [inline tilespec.ts:148]: Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style

File: tilespec.ts [hunk 16]
```diff
@@ -1,54 +1,213 @@
...
+        return Number(coord.replace('~', ''));
+    }
+
+    _parseType(coord: string) {
```
Comment (gonzojive) [inline tilespec.ts:152]: Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style

File: tilespec.ts [hunk 17]
```diff
@@ -1,54 +1,213 @@
...
+        if (coord.includes('~')) {
+            return 'approx_percentage';
+        } else if (coord.includes('.')) {
+            return 'percentage';
+        } else {
+            return 'tile';
+        }
+    }
+
+    _validateTuple(gssk: Array<string>) {
```
Comment (gonzojive) [inline tilespec.ts:162]: Don't use _ for private instance variables, use the private keyword of Typescript: https://google.github.io/styleguide/tsguide.html#naming-style

File: tilespec.ts [hunk 18]
```diff
@@ -83,7 +242,14 @@ function parseGridSizeIgnoringErrors(s: string): GridSize|null {
 }
 
 export class XY {
-    constructor(readonly x: number, readonly y: number) {}
```
Comment (gonzojive) [inline tilespec.ts:86]: This change makes XY mutable, which is undesirable. Do you use the mutability somewhere? If not, please revert. If so, I suggest finding another way.

**Review by gonzojive — COMMENTED** [review]
It would be great if the author could address some of the comments in this review to improve the readability and maintainability of the code base.

File: tilespec.ts [hunk 19]
```diff
@@ -1,54 +1,213 @@
...
-            Math.floor(workArea.size.height / this.gridHeight));
+            workArea.size.width / this.gridWidth,
+            workArea.size.height / this.gridHeight);
+
+        let left;
+        let top;
+        let right;
+        let bottom;
+
+        if (this.luc.types.x == 'tile') {
```
Comment (gonzojive) [inline tilespec.ts:32]: Prefer === over == for comparison in nearly all cases.

File: tilespec.ts [hunk 20]
```diff
@@ -1,54 +1,213 @@
...
+            const positiveTileNumber = this._convertNegativeCoord(this.gridWidth, this.luc.xy.x);
+            left = Math.round(workArea.origin.x + (positiveTileNumber - 1) * elemSize.width);
+        } else if (this.luc.types.x == 'approx_percentage') {
+            const snappedToGrid = Math.round(this.gridWidth * this.luc.xy.x);
+            left = Math.round(workArea.origin.x + snappedToGrid * elemSize.width);
+        } else {
+            left = Math.round(workArea.origin.x + workArea.size.width * this.luc.xy.x);
+        }
+
+        if (this.luc.types.y == 'tile') {
```
Comment (gonzojive) [inline tilespec.ts:42]: Prefer === over == for comparison in nearly all cases.

**Pull request merged**